### DISCUSSION TITLE
[Proposal] Make lock object optional

### DISCUSSION THREAD

**Discussion by rev3r** [dotnet/csharplang#2614] [Language Ideas]
# Background

Everytime we create `lock` block, we need to provide "lock object". It allows declaring multiple `lock` blocks as related. In majority of cases we use it only for single `lock` block.

# Proposal

I suggest making `(_obj)` optional. If not provided, compiler will generate one for us. That will make `lock` cleaner.

```csharp
lock
{
    //code that needs to be synchronized
}
```

# Alternative

Now developers needs to add "lock object" in each statement:

```csharp
private readonly object _obj = new object();
...
lock (_obj)
{
    //code that needs to be synchronized
}
```

That requires more typing and pollutes source code. It also creates possibility of bugs, if not used correctly.

**Comment by HaloFour** [comment]
How would the compiler know where that lock object should be declared?  Should it be instance or static?  How can you share it between multiple methods?   I assume you couldn't.

The clarity of the scope of the lock object is just as important as the lock itself and I don't think the compiler should hide it.

**Comment by rev3r** [comment]
As I said, that would be for 90% cases where you don't need to share "lock object".

How broad scope should be, and if it should be instance or static, could be adjusted to most common scenario.

This is not trying to replace standard `lock`, it's just shortcut for the 90% cases.

**Comment by HaloFour** [comment]
> How broad scope should be, and if it should be instance or static, could be adjusted to most common scenario.

What is that "common scenario"?  What if that changes?

At worst if you're required to declare the lock object you have to type a little extra code once.

**Comment by rev3r** [comment]
@HaloFour most common is `private readonly object _obj = new object();` by far. And if it would change in the future, nobody forces you to ommit `(_obj)`. It only makes life a little easier.
